doc/develop/languages/rust/index.rst

```diff
@@ -0,0 +1,163 @@
+
+Currently, Zephyr supports applications written in Rust and C.  The
+enable Rust support, you must select the :kconfig:option:`CONFIG_RUST`
+in the application configuration file.
+
+The rust toolchain is separate from the rest of the Zephyr SDK.   It
+is recommended to use the `rustup`_ tool to install the rust
+toolchain.  In addition to the base compiler, you will need to install
+core libraries for the target(s) you wish to compile on.  It is
+easiest to determine what needs to be installed by trying a build.
+The diagnostics from the rust compilation will indicate the rust
+command needed to install the appropriate target support:
+
+.. _rustup: https://rustup.rs/
+
+.. code-block::
+
+   $ west build ...
+   ...
+   error[E0463]: can't find crate for `core`
+     |
+   = note: the `thumbv7m-none-eabi` target may not be installed
+   = help: consider downloading the target with `rustup target add thumb7m-none-eabi`
+
+In this case, the given ``rustup`` command will install the needed
+target support.  The target needed will depend on both the board
+selected, as well as certain configuration choices (such as whether
+floating point is enabled).
+
+Writing a Rust Application
+**************************
+
+See :zephyr_file:`samples/rust` for examples of an application.
+
+CMake files
+-----------
+
+The application should contain a :file:`CMakeFiles.txt`, similar to
+the following:
+
+.. code-block:: cmake
+
+   cmake_minimum_required(VERSION 3.20.0)
+
+   find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+   project(hello_world)
+
+   rust_cargo_application()
+
+Cargo files
+-----------
+
+Rust applications are built with Cargo.  The Zephyr build system will
+configure and invoke cargo in your application directory, with options
+set so that it can find the Zephyr support libraries, and that the
+output will be contained within the Zephyr build directory.
+
+The :file:`Cargo.toml` will need to have a ``[lib]`` section that sets
+``crate-type = ["staticlib"]``, and will need to include ``zephyr =
+"0.1.0"`` as a dependency.  You can use crates.io and the Crate
+ecosystem to include any other dependencies you need.  Just make sure
+that you use crates that support building with no-std.
+
+Application
+-----------
+
+The application source itself should live in file:`src/lib.rs`.  A few
+things are needed.  A minimal file would be:
+
+.. code-block:: rust
+
+   #![no_std]
+
+   extern crate zephyr;
+
+   #[no_mangle]
+   extern "C" fn rust_main() {
+   }
+
+The ``no_std`` declaration is needed to prevent the code from
+referencing the ``std`` library.  The extern reference will cause the
+zephyr crate to be brought in, even if nothing from it is used.
+Practically, any meaningful Rust application on Zephyr will use
+something from this crate, and this line is not necessary.  Lastly,
+the main declaration exports the main symbol so that it can be called
+by C code.  The build ``rust_cargo_application()`` cmake function will
```

Comment:
What is the reason for using a C file to call rust main? Is it related to wanting to keep the signature similar to the Rust std main?

Reply:
There isn't a real strong reason to do this. The main issue I ran into is that Zephyr already has a weak symbol for 'main', if it is provided by a library instead of a C source, the linker prefers the weak symbol, and the rust code never gets called.